[CORE-13203] Initial Shadow Link API Implementation

michael-redpanda opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

Adds implementation for the following Shadow Link APIs:

* CreateShadowLink
* GetShadowLink
* ListShadowLinks

Status being returned is not complete, but this is a good start.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

* None

michael-redpanda committed:
resource: Add cluster_link service group

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl: Use explicit namespace for model::node_id

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
app: Wire in cluster_link service

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
utils: Added parse_address to unresolved_address

Adds a static method to net::unresolved_address that will parse a string
and return a net::unresolved_address.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl/errc: Added new error codes

To be used to map cluster::cluster_link::errc responses

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl: Added upsert_link to link_frontend

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl: Add error code mapper

Maps error codes from cluster::cluster_link::errc to cluster_link::errc

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl: Add service handlers

Adds ability to upsert, get and list cluster links to the cluster link
service.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
proto: Minor updates to shadow link

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
admin/sl: Add converter

Add a converter that will convert protobuf requests into
cluster_link::model types.

TODO: Add tests

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
admin/sl: Add mechanism to go from metadata to resource

Now can convert from cluster_link metadata object to shadow link
resource

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
admin/sl: Wire in cluster_link into shadow link services

Implements the create shadow link endpoint

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
admin/sl/tests: Added converter tests

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
dt: Added initial tests of Shadow Linking API

Signed-off-by: Michael Boquard <michael@redpanda.com>

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/71274-->
<!--c9032da0-a70e-4bd2-bca6-d1c424f78085-->
<!--a690c643-4c92-4673-a7e9-90565edc5502-->
<!--042731ff-3296-4bab-8eac-acde13649efa-->
<!--d598cecb-029f-45fc-b629-e54df316d4ec-->
<!--d92ac77b-f158-4d88-9091-7707c0d15222-->
<!--a381383d-a5fc-4050-8f9a-29ef1bb26e0a-->
<!--6b4a65c6-9322-44bb-9b60-f176314eb405-->
<!--82ed3b04-6008-49cb-afd2-67565172221a-->
<!--be688b2c-cc2d-4e78-99a6-131b1db60452-->
<!--title-->
## Retry command for [Build#71274](https://buildkite.com/redpanda/redpanda/builds/71274)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/quick_terminate_test.py::QuickTerminateTest.test_terminate
tests/rptest/tests/scram_test.py::ScramBootstrapUserTest.test_invalid_scram_mechanism@{"expect_fail":true,"mechanism":"sCrAm-ShA-512"}
tests/rptest/tests/node_resize_test.py::NodeResizeTest.test_node_resize
tests/rptest/tests/nodes_decommissioning_test.py::NodesDecommissioningTest.test_node_is_not_allowed_to_join_after_restart@{"new_bootstrap":true}
tests/rptest/tests/nodes_decommissioning_test.py::NodesDecommissioningTest.test_node_is_not_allowed_to_join_after_restart@{"new_bootstrap":false}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--8c91c523-d358-4166-b32d-e65eb91a79c2-->
<!--TEST RESULTS-->
<!--7a2c424d-35be-4b9c-9d99-3d49822688d4-->
<!--TEST RESULTS-->
<!--88bd8aca-74f3-4462-bf41-940d24d74dc7-->
<!--TEST RESULTS-->
<!--271e8333-2cd9-4efb-bd71-55b7b3cbfb8f-->
<!--TEST RESULTS-->
<!--3cda6a30-6e38-4073-9d0b-56dbbaaba304-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/71274">build#71274</a></summary>


| test_class                | test_method                                    | test_arguments                                      | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:--------------------------|:-----------------------------------------------|:----------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTxTest | test_write                                     | null                                                | integration | https://buildkite.com/redpanda/redpanda/builds/71274#0198d3a8-f538-478a-9163-5eea0d5cc4d7 | FLAKY         | 19/21    | upstream reliability is '92.41071428571429'. current run reliability is '90.47619047619048'. drift is 1.93452 and the allowed drift is set to 50. The test should PASS |
| NodeResizeTest            | test_node_resize                               | null                                                | integration | https://buildkite.com/redpanda/redpanda/builds/71274#0198d3a8-f535-4e38-823a-5a6e94439a39 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| NodeResizeTest            | test_node_resize                               | null                                                | integration | https://buildkite.com/redpanda/redpanda/builds/71274#0198d3ad-51e1-4b6f-9222-6e4abaf7320f | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| NodesDecommissioningTest  | test_node_is_not_allowed_to_join_after_restart | {"new_bootstrap": false}                            | integration | https://buildkite.com/redpanda/redpanda/builds/71274#0198d3a8-f536-4629-b438-e366a97b2c30 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| NodesDecommissioningTest  | test_node_is_not_allowed_to_join_after_restart | {"new_bootstrap": false}                            | integration | https://buildkite.com/redpanda/redpanda/builds/71274#0198d3ad-51dd-4972-a80e-fe5846322f7a | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| NodesDecommissioningTest  | test_node_is_not_allowed_to_join_after_restart | {"new_bootstrap": true}                             | integration | https://buildkite.com/redpanda/redpanda/builds/71274#0198d3a8-f537-47ab-a670-0c22707cc7ba | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| NodesDecommissioningTest  | test_node_is_not_allowed_to_join_after_restart | {"new_bootstrap": true}                             | integration | https://buildkite.com/redpanda/redpanda/builds/71274#0198d3ad-51de-4827-a1f2-ff731a765e27 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| QuickTerminateTest        | test_terminate                                 | null                                                | integration | https://buildkite.com/redpanda/redpanda/builds/71274#0198d3a8-f539-451d-b7a3-c38ca85af7cc | FLAKY         | 9/21     | upstream reliability is '100.0'. current run reliability is '42.857142857142854'. drift is 57.14286 and the allowed drift is set to 50. The test should FAIL           |
| ScramBootstrapUserTest    | test_invalid_scram_mechanism                   | {"expect_fail": true, "mechanism": "sCrAm-ShA-512"} | integration | https://buildkite.com/redpanda/redpanda/builds/71274#0198d3a8-f534-4af5-8bb1-7cee76ed34d9 | FAIL          | 0/1      |                                                                                                                                                                        |
| ScramBootstrapUserTest    | test_invalid_scram_mechanism                   | {"expect_fail": true, "mechanism": "sCrAm-ShA-512"} | integration | https://buildkite.com/redpanda/redpanda/builds/71274#0198d3ad-51e0-42a1-aeb4-fe4ee3837008 | FAIL          | 0/1      |                                                                                                                                                                        |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/71337">build#71337</a></summary>


| test_class               | test_method                                    | test_arguments                                      | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                      |
|:-------------------------|:-----------------------------------------------|:----------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------|
| NodeResizeTest           | test_node_resize                               | null                                                | integration | https://buildkite.com/redpanda/redpanda/builds/71337#0198e161-072f-409d-8c8a-5aee472d5ed1 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                      |
| NodeResizeTest           | test_node_resize                               | null                                                | integration | https://buildkite.com/redpanda/redpanda/builds/71337#0198e168-0803-4422-82ea-0ae6f004003a | FAIL          | 0/21     | The test has failed across all retries                                                                                                                      |
| NodesDecommissioningTest | test_node_is_not_allowed_to_join_after_restart | {"new_bootstrap": false}                            | integration | https://buildkite.com/redpanda/redpanda/builds/71337#0198e161-072d-4c79-b06d-f45aab5db19b | FAIL          | 0/21     | The test has failed across all retries                                                                                                                      |
| NodesDecommissioningTest | test_node_is_not_allowed_to_join_after_restart | {"new_bootstrap": false}                            | integration | https://buildkite.com/redpanda/redpanda/builds/71337#0198e168-0800-47d4-95f6-90f988f89526 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                      |
| NodesDecommissioningTest | test_node_is_not_allowed_to_join_after_restart | {"new_bootstrap": true}                             | integration | https://buildkite.com/redpanda/redpanda/builds/71337#0198e161-072e-4a32-b5c0-a93592e468be | FAIL          | 0/21     | The test has failed across all retries                                                                                                                      |
| NodesDecommissioningTest | test_node_is_not_allowed_to_join_after_restart | {"new_bootstrap": true}                             | integration | https://buildkite.com/redpanda/redpanda/builds/71337#0198e168-0801-41ec-a9a9-ae837c8336d1 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                      |
| QuickTerminateTest       | test_terminate                                 | null                                                | integration | https://buildkite.com/redpanda/redpanda/builds/71337#0198e161-072e-461a-9530-1130c3bbdf5b | FLAKY         | 11/21    | upstream reliability is '100.0'. current run reliability is '52.38095238095239'. drift is 47.61905 and the allowed drift is set to 50. The test should PASS |
| QuickTerminateTest       | test_terminate                                 | null                                                | integration | https://buildkite.com/redpanda/redpanda/builds/71337#0198e168-0801-4416-91bd-061275dd1e07 | FLAKY         | 14/21    | upstream reliability is '100.0'. current run reliability is '66.66666666666666'. drift is 33.33333 and the allowed drift is set to 50. The test should PASS |
| ScramBootstrapUserTest   | test_invalid_scram_mechanism                   | {"expect_fail": true, "mechanism": "sCrAm-ShA-512"} | integration | https://buildkite.com/redpanda/redpanda/builds/71337#0198e161-072f-439c-ac40-3eb84c71d4e3 | FAIL          | 0/1      |                                                                                                                                                             |
| ScramBootstrapUserTest   | test_invalid_scram_mechanism                   | {"expect_fail": true, "mechanism": "sCrAm-ShA-512"} | integration | https://buildkite.com/redpanda/redpanda/builds/71337#0198e168-0802-4495-88b9-8031ac04b62f | FAIL          | 0/1      |                                                                                                                                                             |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/71338">build#71338</a></summary>


| test_class                | test_method          | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                |
|:--------------------------|:---------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTxTest | test_write           | null                                                                                                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/71338#0198e209-ea2c-4343-828e-19a73f01f80b | FLAKY         | 19/21    | upstream reliability is '91.5846994535519'. current run reliability is '90.47619047619048'. drift is 1.10851 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest  | test_node_operations | {"cloud_storage_type": 1, "compaction_mode": "adjacent_merge", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/71338#0198e208-0757-4863-a54c-11abdeaf5f2d | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS             |
| RandomNodeOperationsTest  | test_node_operations | {"cloud_storage_type": 2, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}  | integration | https://buildkite.com/redpanda/redpanda/builds/71338#0198e208-0759-407c-8a5a-5766ec7a0bd6 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS             |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/71353">build#71353</a></summary>


| test_class               | test_method          | test_arguments                                                                                                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                     |
|:-------------------------|:---------------------|:---------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/71353#0198e329-46b7-4385-bfef-ba2182ebbd53 | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/71396">build#71396</a></summary>


| test_class                   | test_method        | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-----------------------------|:-------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTxTest    | test_write         | null             | integration | https://buildkite.com/redpanda/redpanda/builds/71396#0198e734-e664-4931-87f8-48e52033aa56 | FLAKY         | 19/21    | upstream reliability is '91.6955017301038'. current run reliability is '90.47619047619048'. drift is 1.21931 and the allowed drift is set to 50. The test should PASS  |
| PartitionMovementUpgradeTest | test_basic_upgrade | null             | integration | https://buildkite.com/redpanda/redpanda/builds/71396#0198e738-417e-47b8-baaf-73dd0b7ba059 | FLAKY         | 19/21    | upstream reliability is '99.45504087193461'. current run reliability is '90.47619047619048'. drift is 8.97885 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

oleiman commented on src/v/cluster_link/manager.cc:
```diff
@@ -112,6 +113,48 @@ ss::future<> manager::stop() {
+manager::create_cluster_link(model::metadata md) {
+    static constexpr auto wait_for_link_creation_timeout = 30s;
+    auto hold = _g.hold();
+    auto name = md.name;
```
nit: what's this copy for?

michael-redpanda replied:
Down a few lines `md` gets moved into `upsert_link` so I keep the name around for logging purposes and to validate that the link gets created

oleiman commented on src/v/cluster_link/manager.cc:
```diff
@@ -112,6 +113,48 @@ ss::future<> manager::stop() {
+            "Aborted waiting for cluster link '{}' to be created", name));
+    }
+
+    auto metadata_resp = _registry->find_link_by_name(name);
```
is it worth checking that the metadata is what we expect, or is the name enough?

michael-redpanda replied:
I guess it's possible that a second upsert comes in and changes things so I don't think we need to validate it

oleiman commented on src/v/cluster_link/manager.cc:
```diff
@@ -112,6 +113,48 @@ ss::future<> manager::stop() {
+    auto name = md.name;
+    vlog(cllog.info, "Attempting to create cluster link named '{}'", md.name);
+    vlog(cllog.trace, "Cluster link metadata: {}", md);
+    auto ec = co_await _registry->upsert_link(
+      std::move(md), ::model::timeout_clock::now() + 30s);
+    auto err = map_cluster_errc(ec);
+    if (err != errc::success) {
+        co_return err_info(
+          err, fmt::format("Failed to create cluster link: {}", ec));
+    }
+
+    try {
+        co_await _link_created_cv.wait(
```
should we limit to one inflight request at a time?

michael-redpanda replied:
That's why I do a name check and only return true if it exists

oleiman replied:
I think this makes a bit more sense to me if we're calling it an upsert 👍

oleiman commented on src/v/utils/unresolved_address.cc:
```diff
@@ -0,0 +1,63 @@
+    auto separator = maybe_address.find(':');
+    if (
+      separator == std::string_view::npos || separator == 0
+      || separator == maybe_address.size() - 1) {
```
suggestion: `std::views::split`?

oleiman commented on tests/rptest/clients/admin/proto/redpanda/core/admin/shadow_link_pb2.py:
```diff
@@ -15,7 +15,7 @@
 from google.protobuf import timestamp_pb2 as google_dot_protobuf_dot_timestamp__pb2
 from google.protobuf import field_mask_pb2 as google_dot_protobuf_dot_field__mask__pb2
-DESCRIPTOR = _descriptor_pool.Default().AddSerializedFile(b'\n+proto/redpanda/core/admin/shadow_link.proto\x12\x13redpanda.core.admin\x1a"proto/redpanda/pbgen/options.proto\x1a\x1eproto/redpanda/pbgen/rpc.proto\x1a$proto/redpanda/core/common/acl.proto\x1a\x1fgoogle/api/field_behavior.proto\x1a\x1bgoogle/api/field_info.proto\x1a\x19google/api/resource.proto\x1a\x1egoogle/protobuf/duration.proto\x1 ...
+DESCRIPTOR = _descriptor_pool.Default().AddSerializedFile(b'\n+proto/redpanda/core/admin/shadow_link.proto\x12\x13redpanda.core.admin\x1a"proto/redpanda/pbgen/options.proto\x1a\x1eproto/redpanda/pbgen/rpc.proto\x1a$proto/redpanda/core/common/acl.proto\x1a\x1fgoogle/api/field_behavior.proto\x1a\x1bgoogle/api/field_info.proto\x1a\x19google/api/resource.proto\x1a\x1egoogle/protobuf/duration.proto\x1 ...
```
the python in this commit is all generated yeah?

michael-redpanda replied:
yessir

oleiman replied:
curious why so much of it changed w/ the tiny proto change, but I guess that's neither here nor there

rockwotj replied:
You don't want to know 😃

mmaslankaprv commented on src/v/utils/unresolved_address.cc:
```diff
@@ -0,0 +1,63 @@
+      separator == std::string_view::npos || separator == 0
+      || separator == maybe_address.size() - 1) {
+        throw std::invalid_argument{fmt_with_ctx(
+          fmt::format,
+          "address override environment variable expected format: 'host:port', "
+          "found: {}",
```
this messages seems to be copied from some other place

mmaslankaprv commented on src/v/utils/unresolved_address.h:
```diff
@@ -53,6 +53,11 @@ class unresolved_address
+    /// \brief Parses a string in the format of "host:port"
+    /// \throws std::invalid_argument if the string is not in the expected
+    /// format
+    static unresolved_address parse_address(std::string_view maybe_address);
```
nit: `from_string` ?

mmaslankaprv replied:
This function would benefit from a unit test

mmaslankaprv commented on src/v/redpanda/admin/services/shadow_link/converter.h:
```diff
@@ -0,0 +1,25 @@
+/// object
+///
+/// \throws std::invalid_argument if the request contains invalid data
+cluster_link::model::metadata
+convert_create_to_metadata(proto::admin::create_shadow_link_request req);
```
test ?

michael-redpanda replied:
I added unit tests in a separate commit

mmaslankaprv commented on src/v/cluster_link/manager.h:
```diff
@@ -55,6 +55,7 @@ class manager {
     ss::future<> stop();
 
     ss::future<result<model::metadata>> create_cluster_link(model::metadata md);
+    result<model::metadata> get_cluster_link(model::name_t name);
```
do we need a copy of name ?

rockwotj replied:
nit: docs

michael-redpanda commented:
Force push:

* Addressed PR comments

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/71337-->
<!--af65c229-58c5-4eb3-9d6e-852788171f89-->
<!--c3351e10-e0a8-484d-acac-3944e795f85a-->
<!--5a0f5770-7f38-42f3-80c3-c095c8b104c4-->
<!--837e145c-0f8e-4790-935f-f1e237cbf3b1-->
<!--59de2e06-2562-40d7-869e-93df563eec5f-->
<!--fcdfe363-6aea-4b5d-9fd4-2727685292b9-->
<!--65f8738f-21d6-47af-8cdd-a379c19440a8-->
<!--20ceb376-eb97-4935-9e90-aaf8efcff881-->
<!--title-->
## Retry command for [Build#71337](https://buildkite.com/redpanda/redpanda/builds/71337)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/scram_test.py::ScramBootstrapUserTest.test_invalid_scram_mechanism@{"expect_fail":true,"mechanism":"sCrAm-ShA-512"}
tests/rptest/tests/nodes_decommissioning_test.py::NodesDecommissioningTest.test_node_is_not_allowed_to_join_after_restart@{"new_bootstrap":true}
tests/rptest/tests/node_resize_test.py::NodeResizeTest.test_node_resize
tests/rptest/tests/nodes_decommissioning_test.py::NodesDecommissioningTest.test_node_is_not_allowed_to_join_after_restart@{"new_bootstrap":false}
```
<!--/body-->

rockwotj commented on src/v/cluster_link/service.h:
```diff
@@ -54,7 +54,7 @@ class service {
 private:
     ss::gate _gate;
-    model::node_id _self;
+    ::model::node_id _self;
```
A reason why this is done would be useful in the commit message

rockwotj commented on src/v/cluster_link/service.h:
```diff
@@ -47,6 +49,7 @@ class service {
 
     ss::future<> start();
     ss::future<> stop();
+    ss::future<result<model::metadata>> create_cluster_link(model::metadata md);
```
Please add documentation to this method.
Should this be called `upsert`. Create should fail if the link already exists.

rockwotj commented on src/v/cluster_link/manager.cc:
```diff
@@ -112,6 +113,48 @@ ss::future<> manager::stop() {
+          errc::link_id_not_found,
+          fmt::format("Failed to find cluster link with name '{}'", name));
+    }
+
```
as a user calling `create_cluster_link` it makes no sense to get back a not found error.

I am also confused as to why a broken condition variable means that cluster linking is disabled - isn't it for shutdown instead?

rockwotj commented on src/v/utils/unresolved_address.cc:
```diff
@@ -0,0 +1,49 @@
+    auto host_view = maybe_address.substr(0, separator);
+    auto port_view = maybe_address.substr(separator + 1);
+
+    uint16_t port = 0;
+    auto result = std::from_chars(
+      port_view.data(), port_view.data() + port_view.size(), port);
+
+    if (result.ec != std::errc{}) {
+        throw std::invalid_argument{fmt_with_ctx(
+          fmt::format, "failed to convert {} to port (uint16_t)", port_view)};
+    }
+
+    if (result.ptr != port_view.data() + port_view.size()) {
+        throw std::invalid_argument{fmt_with_ctx(
+          fmt::format, "failed to convert {} to port (uint16_t)", port_view)};
+    }
```
`absl::SimpleAtoi<uint16_t>` is much simpler :)

michael-redpanda replied:
Actually only works for 32-bit or 64-bit numbers

michael-redpanda replied:
But I'll switch to using it

mmaslankaprv commented:
Tests fails with segmentation faults which seems to be related.

rockwotj commented on src/v/redpanda/admin/services/shadow_link/converter.cc:
```diff
@@ -0,0 +1,266 @@
+/*
+ * Copyright 2025 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
```
nit: shouldn't all shadow link stuff be enterprised licensed?

rockwotj commented on src/v/redpanda/admin/services/shadow_link/converter.cc:
```diff
@@ -0,0 +1,266 @@
+
+/// \brief Converts protobuf scram_mechanism to string
+/// \throws std::invalid_argument if the mechanism is unspecified
+constexpr auto scram_mechanism_to_string(scram_mechanism m)
+  -> std::string_view {
```
why trailing return syntax?

rockwotj commented on src/v/redpanda/admin/services/shadow_link/converter.cc:
```diff
@@ -0,0 +1,266 @@
+        throw serde::pb::rpc::invalid_argument_exception(
+          ssx::sformat("Invalid cluster link configuration: {}", e.what()));
+    } catch (const std::exception& e) {
+        throw serde::pb::rpc::unknown_exception(
+          ssx::sformat(
+            "Failed to convert create cluster link request: {}", e.what()));
```
should we just let these bubble up? Then the handler can convert this into an `internal` error.

michael-redpanda replied:
You mean throw something else and then convert to a `serde::pb::rpc` error?  As this is in the admin services namespace I feel the way I did it was appropriate.

rockwotj replied:
This is OK, but this is a catch all for unexpected errors right? That should be `internal` not `unknown`

rockwotj replied:
If it's a catch all, I think it's better to let it bubble up and get logged like other unexpected errors, but it's OK to do here, especially if we want to do something different, but I don't see what we want to differently?

michael-redpanda replied:
> If it's a catch all, I think it's better to let it bubble up and get logged like other unexpected errors,

You mean let the RPC layer handle that?  The layer above my service implementation?

rockwotj replied:
yeah

rockwotj commented on src/v/redpanda/admin/services/shadow_link/converter.h:
```diff
@@ -0,0 +1,25 @@
+/// \brief Converts a create cluster link request into a cluster link metadata
+/// object
+///
+/// \throws std::invalid_argument if the request contains invalid data
```
needs updating

rockwotj commented on src/v/redpanda/admin/services/shadow_link/converter.cc:
```diff
@@ -0,0 +1,266 @@
+        -> cluster_link::model::connection_config::authn_variant {
+          cluster_link::model::scram_credentials creds;
+          creds.username = scram.get_username();
+          creds.password = scram.get_password();
```
will something else enforce the password is not empty string?

michael-redpanda replied:
Currently nothing - I'll add it to the frontend validator

rockwotj commented on src/v/redpanda/application.cc:
```diff
@@ -1504,6 +1505,23 @@ void application::wire_up_runtime_services(
           .get();
     }
 
+    construct_service(
+      _cluster_link_service,
+      node_id,
+      &controller->get_cluster_link_frontend(),
+      ss::sharded_parameter([this] {
+          return cluster::partition_change_notifier_impl::make_default(
+            raft_group_manager,
+            partition_manager,
+            controller->get_topics_state());
+      }),
+      &partition_manager,
+      &controller->get_partition_leaders(),
+      &controller->get_shard_table(),
+      &metadata_cache,
+      smp_service_groups.cluster_link_smp_sg())
+      .get();
```
does this need to be behind a feature flag?

michael-redpanda replied:
I have the table updates behind the feature flag (or at this point an experimental cluster config, which will convert to a feature flag soon)

rockwotj replied:
Is there any risk to constructing this service but leaving the feature flag be low level? Our general pattern is to stub it out at this level which is kind of the "safest" likely because of making sure we don't have some bug that would effect clusters with this feature unused (think some bug in controller log that causes more to be read or the snapshot breaks something on old clusters, etc). If there is no risk then it's OK IMO to leave this unflagged, but it is different from what we normally do.

michael-redpanda replied:
Based on your example provided, I feel like it's safe because we don't even attempt to write to the controller log if it isn't enabled.  I mean I can do the checks in the API service as well if that makes things more understandable.

rockwotj replied:
I'm not fully opposed, it's just different than our normal practice.

rockwotj commented on src/v/redpanda/admin/services/shadow_link/shadow_link.cc:
```diff
@@ -11,13 +11,79 @@
+    case cluster_link::errc::failed_to_connect_to_remote_cluster:
+    case cluster_link::errc::remote_cluster_does_not_support_required_api:
+    case cluster_link::errc::link_connection_failed:
+    case cluster_link::errc::cluster_link_disabled:
```
shouldn't disabled be a `failed_precondition`

rockwotj commented on src/v/redpanda/admin/services/shadow_link/shadow_link.cc:
```diff
@@ -11,13 +11,79 @@
+    case cluster_link::errc::rpc_error:
+        throw serde::pb::rpc::internal_exception(info.message());
+    case cluster_link::errc::failed_to_connect_to_remote_cluster:
+    case cluster_link::errc::remote_cluster_does_not_support_required_api:
```
seems like a `invalid_argument`?

michael-redpanda replied:
I don't feel like these are 4xx errors.  Like everything user could have parsed correctly and the target cluster just isn't ready or at the right version, which feels more like a 5xx sort of error.

rockwotj commented on src/v/redpanda/admin/services/shadow_link/shadow_link.cc:
```diff
@@ -11,13 +11,79 @@
+    case cluster_link::errc::failed_to_start_task:
+    case cluster_link::errc::task_already_registered_on_link:
+    case cluster_link::errc::task_creation_failed:
+    case cluster_link::errc::rpc_error:
```
`unavailable_exception`?

michael-redpanda replied:
I don't think so - it's not necessariliy that the service is unavailable but some internal process/action failed.

rockwotj commented on src/v/redpanda/admin/services/shadow_link/shadow_link.cc:
```diff
@@ -11,13 +11,79 @@
+    return parser.read_string_unsafe(parser.bytes_left());
+}
+
+template<typename T>
+ss::future<ss::sstring> request_to_json_string(const T& req) {
+    return req.to_json().then(
+      [](iobuf buf) { return iobuf_to_string(std::move(buf)); });
+}
+} // namespace
```
no thank you please use formatters

michael-redpanda replied:
ha - sorry i was debugging something and this stayed in

rockwotj commented on src/v/redpanda/admin/services/shadow_link/shadow_link.cc:
```diff
@@ -11,13 +11,79 @@
+
+    auto md = convert_create_to_metadata(std::move(req));
+    auto resp = handle_error(
+      co_await _service->local().create_cluster_link(std::move(md)));
```
do we need to handle the sharded service being disabled because the feature flag is off?

rockwotj commented on tests/rptest/tests/cluster_linking_e2e_test.py:
```diff
@@ -117,3 +123,141 @@ def test_basic_ops(self):
+        return resp.message().shadow_link
+
+    @cluster(num_nodes=6)
+    def test_service_exists(self):
```
hopefully overtime we can remove this and it's implicit because we test the behavior of each endpoint and not the existance of them?

michael-redpanda replied:
Yeah I'm happy to just pull this out - I was testing/trying things out

rockwotj replied:
up for nuking now that we're writing real tests

rockwotj commented on tests/rptest/tests/cluster_linking_e2e_test.py:
```diff
@@ -117,3 +123,141 @@ def test_basic_ops(self):
+    def list_all_shadow_links(
+        self, client: shadow_link_pb2_connect.ShadowLinkServiceClient
+    ) -> list[shadow_link_pb2.ShadowLink]:
+        resp = client.call_list_shadow_links(
+            req=shadow_link_pb2.ListShadowLinksRequest())
+        if resp.error() is not None:
+            raise resp.error()
```
just call `list_shadow_links`?

rockwotj commented on tests/rptest/tests/cluster_linking_e2e_test.py:
```diff
@@ -117,3 +123,141 @@ def test_basic_ops(self):
+            shadow_link_configurations)
+        create_shadow_link_request.shadow_link.CopyFrom(shadow_link_resource)
+
+        shadow_link_create_result = shadow_link.call_create_shadow_link(
```
Why use `call_` version and instead just let it throw?

rockwotj commented on tests/rptest/tests/cluster_linking_e2e_test.py:
```diff
@@ -117,3 +123,141 @@ def test_basic_ops(self):
+    def get_shadow_link(
+            self, client: shadow_link_pb2_connect.ShadowLinkServiceClient,
+            name: str) -> shadow_link_pb2.ShadowLink:
+        resp = client.call_get_shadow_link(
```
why use the `call_` version

michael-redpanda commented:
Force push:

* Fixed segmentation fault in cluster_link::service when service stops before it fully comes up and `manager` hasn't been created yet

oleiman commented on src/v/redpanda/admin/services/shadow_link/shadow_link.cc:
```diff
@@ -94,8 +94,17 @@ shadow_link_service_impl::delete_shadow_link(
-    throw serde::pb::rpc::unimplemented_exception();
+  proto::admin::get_shadow_link_request req) {
+    if (sllog.is_enabled(ss::log_level::trace)) {
+        auto json_str = co_await request_to_json_string(req);
+        vlog(sllog.trace, "get_shadow_link: {}", json_str);
```
why do we need this? is the request not formattable?

michael-redpanda replied:
it gone

oleiman commented on src/v/cluster_link/manager.cc:
```diff
@@ -165,6 +165,24 @@ result<model::metadata> manager::get_cluster_link(const model::name_t& name) {
+
+    for (const auto id : link_ids) {
+        auto maybe_md = _registry->find_link_by_id(id);
+        if (!maybe_md) {
+            vlog(cllog.warn, "Failed to find link ID {}", id);
+            continue;
+        }
```
q: would it make sense to include this in the response in some form? put another way, is an outstanding link ID w/ no associated metadata something a client should know about?

michael-redpanda replied:
I actually don't think this can happen, `get_all_link_ids` is built from the internal table which is key'ed by ID

oleiman replied:
should we enforce that as an invariant then? dumb example like if the implementation of `get_all_link_ids` changed to use some caching or something?

oleiman commented on src/v/cluster_link/manager.h:
```diff
@@ -55,7 +55,8 @@ class manager {
 
     ss::future<result<model::metadata>> create_cluster_link(model::metadata md);
-    result<model::metadata> get_cluster_link(model::name_t name);
+    result<model::metadata> get_cluster_link(const model::name_t& name);
```
i think this belongs in the previous commit

michael-redpanda committed:
cl/fe: Validating SCRAM configuration

Ensures SCRAM configuration is valid - no empty username, no empty
password, and provided mechanism is valid.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda commented:
Force push:

* Addressed comments and lint error

michael-redpanda commented:
Force push:

* Rebased

rockwotj reviewed (approved):
LGTM. I probably need to double check look at the converter logic again closer.
